### PR TITLE
ncm-ssh: add ssh/sshd option GSSAPIKexAlgorithms to schema

### LINKED ISSUE #1713: ncm-ssh: add GSSAPIKexAlgorithms to PAN schema.
Ensure the title of this issue starts with the component name followed by a colon, e.g.
> ncm-ssh should include ssh/sshd config option GSSAPIKexAlgorithms in the PAN schema.

* The version of the component you are using (rpm -q $component) 23.6.1
* What OS version you are using rhel6,7,8,9
* Steps required to reproduce the problem
* Actual result versus expected result

### PR DESCRIPTION
**Pull request by ulrich1919**
Fixes #1713

### CODE REVIEW

File: ncm-ssh/src/main/pan/components/ssh/schema.pan [hunk 1]
```diff
@@ -73,6 +73,7 @@ type ssh_core_options_type = {
     "Compression" ? string with match (SELF, '^(yes|delayed|no)$')
     "GSSAPIAuthentication" ? legacy_binary_affirmation_string
     "GSSAPICleanupCredentials" ? legacy_binary_affirmation_string
+    "GSSAPIKexAlgorithms" ? string
```
Comment (jrha) [inline ncm-ssh/src/main/pan/components/ssh/schema.pan:76]: This should at the very least be `string_trimmed`, but presumably this is a comma separated list like `KexAlgorithms`?

`man` suggests the following values are valid?
```
gss-gex-sha1-
gss-group1-sha1-
gss-group14-sha1-
gss-group14-sha256-
gss-group16-sha512-
gss-nistp256-sha256-
gss-curve25519-sha256-
```

Comment (ulrich1919) [reply ncm-ssh/src/main/pan/components/ssh/schema.pan]: makes sense, amended

File: ncm-ssh/src/main/pan/components/ssh/schema.pan [hunk 2]
```diff
@@ -24,6 +24,8 @@ type ssh_kbdinteractivedevices = string with match (SELF, "^(bsdauth|pam|skey)$"
 type ssh_kexalgorithms = string with match (SELF, "^(diffie-hellman-group-exchange-sha256|" +
     "ecdh-sha2-nistp(256|384|521)|curve25519-sha256@libssh.org)$");
 type ssh_MACs = string with is_valid_ssh_MAC(SELF);
+type ssh_gssapikexalgorithms = string with match (SELF, "^(gss-gex-sha1-|gss-group1-sha1-|" +
+    "gss-group14-sha1-|gss-group14-sha256-|gss-group16-sha512-|gss-nistp256-sha256-|gss-curve25519-sha256-)$");
```
Comment (jrha) [inline ncm-ssh/src/main/pan/components/ssh/schema.pan:28]: Thanks, `choice()` is the more modern solution for this, but this fits in with the rest of the schema so it's fine.

File: ncm-ssh/src/main/pan/components/ssh/schema.pan [hunk 3]
```diff
@@ -73,6 +75,7 @@ type ssh_core_options_type = {
     "Compression" ? string with match (SELF, '^(yes|delayed|no)$')
     "GSSAPIAuthentication" ? legacy_binary_affirmation_string
     "GSSAPICleanupCredentials" ? legacy_binary_affirmation_string
+    "GSSAPIKexAlgorithms" ? ssh_gssapikexalgorithms[]
```
Comment (jrha) [inline ncm-ssh/src/main/pan/components/ssh/schema.pan:78]: I think this should be `ssh_gssapikexalgorithms[1..]` as a missing value causes the component to fail at run-time.

**Comment by ulrich1919** [comment]
oh, btw choice() is a single choice right?
we need, multiple algorithms to be added. So choice() is not suitable here.

**Comment by jrha** [comment]
It's a list of items from `choice()` though so you can still select multiple.

File: ncm-ssh/src/main/pan/components/ssh/schema.pan [hunk 4]
```diff
@@ -73,6 +73,15 @@ type ssh_core_options_type = {
...
     "GSSAPICleanupCredentials" ? legacy_binary_affirmation_string
+    "GSSAPIKexAlgorithms" ? choice(
+        'gss-gex-sha1-',
+        'gss-group1-sha1-',
+        'gss-group14-sha1-',
+        'gss-group14-sha256-',
+        'gss-group16-sha512-',
+        'gss-nistp256-sha256-',
+        'gss-curve25519-sha256-'
+    )
```
Comment (jrha) [inline ncm-ssh/src/main/pan/components/ssh/schema.pan:84]: Restore this:
```suggestion
    "GSSAPIKexAlgorithms" ? ssh_gssapikexalgorithms[1..]
```

and add
```pan
    type ssh_gssapikexalgorithms = ssh_gssapikexalgorithms` ? choice(
        'gss-gex-sha1-',
        'gss-group1-sha1-',
        'gss-group14-sha1-',
        'gss-group14-sha256-',
        'gss-group16-sha512-',
        'gss-nistp256-sha256-',
        'gss-curve25519-sha256-'
    )

**Review by jrha — CHANGES_REQUESTED** [review]

File: ncm-ssh/src/main/pan/components/ssh/schema.pan [hunk 5]
```diff
@@ -24,6 +24,15 @@ type ssh_kbdinteractivedevices = string with match (SELF, "^(bsdauth|pam|skey)$"
...
 type ssh_MACs = string with is_valid_ssh_MAC(SELF);
+type ssh_gssapikexalgorithms = choice(
+        'gss-gex-sha1-',
+        'gss-group1-sha1-',
+        'gss-group14-sha1-',
+        'gss-group14-sha256-',
+        'gss-group16-sha512-',
+        'gss-nistp256-sha256-',
+        'gss-curve25519-sha256-'
+    )
```
Comment (jrha) [inline ncm-ssh/src/main/pan/components/ssh/schema.pan:35]: Missing `;` here.

**Review by jrha — APPROVED** [review]

**Pull request merged**